### DISCUSSION TITLE
AES Source Documentation

### DISCUSSION THREAD

**Discussion by wdkaye** [DominoJP/Image-Encryption-AES#3] [General]
Clicking through references on Wikipedia, I found the original source documentation for AES, a doc called "[FIPS 197](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.197-upd1.pdf)".

Section 5 presents both the encryption (5.1) and decryption (5.3) algorithms in pseudocode.  It's gnarly to look at, but sections 1-4 explain the acronyms and symbols, and how to do the math.

**Comment by Andrew-Miner** [comment]
I just noticed the GeeksForGeeks link I posted doesn't talk about the key expansion process at all. This original document is way better.

**Reply by Andrew-Miner** [reply]
All the math jargon makes it seems way more complicated than it is.

**Reply by wdkaye** [reply]
Agreed.  Sections 1-4 do a good job of breaking down the fundamental ideas.

**Comment by Andrew-Miner** [comment]
One important thing to note for the key expansion. The table for RCON in section 5.2 starts at index 1 and not 0. So in the pseudo code below it you need to change:
```C++
 Rcon[i/Nk]
```
to
```C++
 Rcon[i/Nk - 1]
```

I spent way too long debugging that.

**Comment by Andrew-Miner** [comment]
Another really important thing to note is that the MixColumns matrix multiplication step is not normal matrix multiplication. The additions are really Xors and the multiplication is "performed over a finite field (also known as Galois Field GF(2^8))". If you ask ChatGPT about the MixColumn multiplication step it will give you a decent breakdown of how to do it. Since the constant matrix we use only has 1, 2, and 3 as values we can create a special case for each.

**Reply by Andrew-Miner** [reply]
Here's what I came up with using ChatGPT for some explanations:
```C++


unsigned char galoisMultiplyBy2(unsigned char value) {
    unsigned char result = value << 1;
    if (value & 0x80) { // If the most significant bit is set (overflow)
        result ^= 0x1b; // XOR with the AES irreducible polynomial
    }
    return result;
}

std::vector<unsigned char> mixColumns(std::vector<unsigned char>& buffer, const int rowCount)
{
    static const unsigned char COL_MIXER[AES_BLOCK_COLS][AES_BLOCK_ROWS] = {
        {0x02, 0x03, 0x01, 0x01},
        {0x01, 0x02, 0x03, 0x01},
        {0x01, 0x01, 0x02, 0x03},
        {0x03, 0x01, 0x01, 0x02},
    };

    assert(buffer.size() % rowCount == 0);

    std::vector<unsigned char> mixed = std::vector<unsigned char>(BUFFER_SIZE, 0);
    int colCount = buffer.size() / rowCount;
    for (int col = 0; col < colCount; ++col) {

        for (int mixerRow = 0; mixerRow < AES_BLOCK_ROWS; ++mixerRow) {
            unsigned char mixedValue = 0;  // Temporary value to accumulate results
            for (int mixerCol = 0; mixerCol < AES_BLOCK_COLS; ++mixerCol) {
                unsigned char temp = 0;
                unsigned char value = buffer[col * rowCount + mixerCol];
                switch (COL_MIXER[mixerRow][mixerCol]) {
                    case 1:
                        temp = value;
                        break;
                    case 2:
                        temp = galoisMultiplyBy2(value);
                        break;
                    case 3:
                        temp = galoisMultiplyBy2(value) ^ value;
                        break;
                    default:
                        std::cout << "Error: Invalid Constant Array!" << std::endl;
                        return std::vector<unsigned char>();
                }
                mixedValue ^= temp;
            }
            mixed[col * rowCount + mixerRow] = mixedValue;
        }

    }

    return mixed;
}
```
Can confirm it works.
